x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/Base64EncodeToString.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.scalar.convert;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.compute.ann.ConvertEvaluator;
+import org.elasticsearch.xpack.esql.evaluator.mapper.EvaluatorMapper;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.ql.type.DataTypes.KEYWORD;
+import static org.elasticsearch.xpack.ql.type.DataTypes.TEXT;
+
+public class Base64EncodeToString extends AbstractConvertFunction implements EvaluatorMapper {
+
+    private static final Map<DataType, BuildFactory> EVALUATORS = Map.ofEntries(
+        Map.entry(KEYWORD, Base64EncodeToStringEvaluator.Factory::new),
+        Map.entry(TEXT, Base64EncodeToStringEvaluator.Factory::new)
+    );
+
+    @FunctionInfo(
+        returnType = "keyword",
+        description = "Encode a string to a base64 string.",
+        examples = @Example(file = "string", tag = "base64_encode")
+    )
+    public Base64EncodeToString(
+        Source source,
+        @Param(name = "string", type = { "keyword", "text" }, description = "A string.") Expression string
+    ) {
+        super(source, string);
+    }
+
+    @Override
+    protected Map<DataType, BuildFactory> factories() {
+        return EVALUATORS;
+    }
+
+    @Override
+    public DataType dataType() {
+        return KEYWORD;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Base64EncodeToString(source(), newChildren.get(0));
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Base64EncodeToString::new, field());
+    }
+
+    @ConvertEvaluator()
+    static BytesRef process(BytesRef input) {
+        byte[] bytes = new byte[input.length];
+        System.arraycopy(input.bytes, input.offset, bytes, 0, input.length);
+        return new BytesRef(Base64.getEncoder().encode(bytes));
```

Comment:
It's unfortunate internally Base64 creates an array - can we avoid it on our end at least and reuse the created array (`bytes`) for a given block?

Reply:
Originally, I use BytesRef.bytes as input to Base64's encode/decode methods, but I got some garbage, then I realized that BytesRef has offset and length too, it is not guaranteed that the bytes.length equals to BytesRef.length or offset is 0, will see if there is a way to avoid allocating new storage.

Reply:
> It's unfortunate internally Base64 creates an array - can we avoid it on our end at least and reuse the created array (`bytes`) for a given block?

The source to encode/decode is trickier than the target, the Base64 encoder/decoder from the jdk expects the length of the source array to be exactly the same as the string to be encoded/decoded, otherwise it will pad "A" or "=" for the extra bytes, BytesRef always allocates extra storage, its length is much larger than the string, the jdk has another method decode0, this is what we need as it takes the source array, a length and target array as inputs, however it is private. If BytesRefBuilder can shrink, it might works for the source, however it only grows...